### PR TITLE
Fix/staking epoch finalization

### PR DESCRIPTION
**Pull request by mzhu25**
## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: contracts/staking/contracts/src/Staking.sol [hunk 1]
```diff
@@ -38,9 +38,18 @@ contract Staking is
...
+        // Patch corrupted state
+        aggregatedStatsByEpoch[prevEpoch].numPoolsToFinalize = 0;
+        this.endEpoch();
+
+        uint256 lastPoolId_ = uint256(lastPoolId);
+        for (uint256 i = 1; i <= lastPoolId_; i++) {
+            this.finalizePool(bytes32(i));
+        }
+        // Ensure that current epoch's state is not corrupted
+        aggregatedStatsByEpoch[currentEpoch_].numPoolsToFinalize = 0;
```
Comment (dorothy-zbornak) [inline contracts/staking/contracts/src/Staking.sol:53]: I think we shouldn't merge this. It'll probably break stuff also?

File: contracts/staking/test/patch_mainnet_test.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        const staking = new StakingContract(STAKING_PROXY, env.provider, undefined, abis);
+        const maker = '0x7b1886e49ab5433bb46f7258548092dc8cdca28b';
+        const zeroFeeTx = await staking
+            .payProtocolFee(maker, constants.NULL_ADDRESS, constants.ZERO_AMOUNT)
+            .awaitTransactionSuccessAsync({ from: EXCHANGE_PROXY, gasPrice: 0 }, { shouldValidate: false });
+        // StakingPoolEarnedRewardsInEpoch should _not_ be emitted for a zero protocol fee.
+        // tslint:disable-next-line:no-unused-expression
+        expect(filterLogsToArguments(zeroFeeTx.logs, StakingEvents.StakingPoolEarnedRewardsInEpoch)).to.be.empty;
+
+        // Coincidentally there's some ETH in the ExchangeProxy
```
Comment (dorothy-zbornak) [inline contracts/staking/test/patch_mainnet_test.ts:56]: lol "some"

**Review by dorothy-zbornak — APPROVED** [review]

File: contracts/staking/contracts/src/Staking.sol (hunk 1 above)
Comment (mzhu25) [reply contracts/staking/contracts/src/Staking.sol:53]: I have this here in case the same situation arises this epoch before this fix goes in

**Review by hysz — APPROVED** [review]
Looks good to me!

**Pull request merged**